diff --git a/include/linux/mdcast.h b/include/linux/mdcast.h
new file mode 100644
index 000000000000..c80a93756c64
--- /dev/null
+++ b/include/linux/mdcast.h
@@ -0,0 +1,64 @@
+#ifndef _LINUX_MDCAST_H
+#define _LINUX_MDCAST_H
+
+#include "linux/in6.h"
+#include "linux/ipv6.h"
+#include "linux/stddef.h"
+#include <linux/ipv6.h>
+#include <linux/skbuff.h>
+
+int md_rcv(struct sk_buff *skb);
+int md_dcv_rcv(struct sk_buff *skb);
+int dcv_rsp(struct sk_buff *skb);
+
+/*
+ * May using similar approach as rpl.h::ipv6_rpl_sr_hdr.
+ * For example, segments property is used in exthdrs.c:585
+ */
+
+/*
+ * Fix sized part of MEADcast header, bc so far I'm not sure how to handle two
+ * var sized properties in struct (dst array and router map)
+ * substitute hop-count by segments left
+ *  8 bit num dsts
+ *  1 bit disc flag
+ *  1 bit resp flag
+ * 30 bit reserved
+ */
+
+// struct rtmd_hdr {
+// 	struct ipv6_rt_hdr	rt_hdr;
+//     __u8    num_dst;
+//     int dicover: 1;
+//     int response: 1;
+//     int reserved: 30;
+//     __u32 dlvmap;
+//     __u32 rtmap;
+//     union {
+//         __DECLARE_FLEX_ARRAY(struct in6_addr, addr);
+//         __DECLARE_FLEX_ARRAY(__u16, ports);
+//     } addresses;
+//
+// #define rtmd_dstaddr addresses.addr
+// #define rtmd_dstrtmap addresses.ports
+// #define rtmd_type		rt_hdr.type
+// };
+
+struct ipv6_rt_md_hdr {
+    struct ipv6_rt_hdr rt_hdr;
+    __u8 num_dst;
+    __u32
+        discovery:  1,
+        response:   1,
+        hopcount:   6,
+        reserved:  24;
+    __u32 dlvmap;
+    __u32 rtmap;
+    struct in6_addr dst[];
+
+#define rt_md_type		rt_hdr.type
+} __attribute__((packed));
+
+void print_debug(struct ipv6hdr *iphdr, struct ipv6_rt_md_hdr *hdr);
+
+#endif // !_LINUX_MDCAST_H
diff --git a/include/uapi/linux/ipv6.h b/include/uapi/linux/ipv6.h
index ac56605fe9bc..55d0fe29f445 100644
--- a/include/uapi/linux/ipv6.h
+++ b/include/uapi/linux/ipv6.h
@@ -43,6 +43,9 @@ struct in6_ifreq {
 #define IPV6_SRCRT_TYPE_2	2	/* IPv6 type 2 Routing Header	*/
 #define IPV6_SRCRT_TYPE_3	3	/* RPL Segment Routing with IPv6 */
 #define IPV6_SRCRT_TYPE_4	4	/* Segment Routing with IPv6 */
+#ifdef CONFIG_IPV6_MEADCAST
+#define IPV6_MEADCAST 253   /* Using experimental header for MEADcast */
+#endif
 
 /*
  *	routing header
diff --git a/net/ipv6/Kconfig b/net/ipv6/Kconfig
index 08d4b7132d4c..809bd04a3476 100644
--- a/net/ipv6/Kconfig
+++ b/net/ipv6/Kconfig
@@ -340,4 +340,13 @@ config IPV6_IOAM6_LWTUNNEL
 
 	  If unsure, say N.
 
+config IPV6_MEADCAST
+	bool "IPv6: MEADcast support"
+	depends on IPV6
+    default y
+	help
+	  Support for MEADcast (Multicast to Explicit Agnostic Destinations).
+
+	  If unsure, say N.
+
 endif # IPV6
diff --git a/net/ipv6/Makefile b/net/ipv6/Makefile
index 3036a45e8a1e..9a45c28b69ec 100644
--- a/net/ipv6/Makefile
+++ b/net/ipv6/Makefile
@@ -26,6 +26,7 @@ ipv6-$(CONFIG_IPV6_SEG6_LWTUNNEL) += seg6_iptunnel.o seg6_local.o
 ipv6-$(CONFIG_IPV6_SEG6_HMAC) += seg6_hmac.o
 ipv6-$(CONFIG_IPV6_RPL_LWTUNNEL) += rpl_iptunnel.o
 ipv6-$(CONFIG_IPV6_IOAM6_LWTUNNEL) += ioam6_iptunnel.o
+ipv6-$(CONFIG_IPV6_MEADCAST) += mdcast.o
 
 obj-$(CONFIG_INET6_AH) += ah6.o
 obj-$(CONFIG_INET6_ESP) += esp6.o
diff --git a/net/ipv6/exthdrs.c b/net/ipv6/exthdrs.c
index 202fc3aaa83c..dd0338174908 100644
--- a/net/ipv6/exthdrs.c
+++ b/net/ipv6/exthdrs.c
@@ -28,6 +28,10 @@
 #include <linux/slab.h>
 #include <linux/export.h>
 
+#ifdef CONFIG_IPV6_MEADCAST
+#include <linux/mdcast.h>
+#endif /* ifdef CONFIG_IPV6_MEADCAST */
+
 #include <net/dst.h>
 #include <net/sock.h>
 #include <net/snmp.h>
@@ -684,6 +688,10 @@ static int ipv6_rthdr_rcv(struct sk_buff *skb)
 	case IPV6_SRCRT_TYPE_3:
 		/* rpl segment routing */
 		return ipv6_rpl_srh_rcv(skb);
+#ifdef CONFIG_IPV6_MEADCAST
+    case IPV6_MEADCAST:
+        return md_rcv(skb);
+#endif
 	default:
 		break;
 	}
diff --git a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
index 1e8c90e97608..ddcdad1c740c 100644
--- a/net/ipv6/ip6_output.c
+++ b/net/ipv6/ip6_output.c
@@ -22,6 +22,7 @@
  *				for datagram xmit
  */
 
+#include "linux/mdcast.h"
 #include <linux/errno.h>
 #include <linux/kernel.h>
 #include <linux/string.h>
@@ -647,6 +648,10 @@ int ip6_forward(struct sk_buff *skb)
 
 	hdr->hop_limit--;
 
+    if (hdr->nexthdr == IPPROTO_ROUTING) {
+        md_dcv_rcv(skb);
+    }
+
 	return NF_HOOK(NFPROTO_IPV6, NF_INET_FORWARD,
 		       net, NULL, skb, skb->dev, dst->dev,
 		       ip6_forward_finish);
diff --git a/net/ipv6/mdcast.c b/net/ipv6/mdcast.c
new file mode 100644
index 000000000000..c2908abedf96
--- /dev/null
+++ b/net/ipv6/mdcast.c
@@ -0,0 +1,147 @@
+#include "linux/ipv6.h"
+#include "linux/skbuff.h"
+#include "net/dst.h"
+#include "net/ip6_route.h"
+#include <linux/printk.h>
+#include <linux/mdcast.h>
+#include <linux/in.h>
+#include <linux/string.h>
+
+#ifndef INET6_ADDRSTRLEN
+#define INET6_ADDRSTRLEN 46
+#endif
+
+#ifndef MEADCAST_DEBUG
+#define MEADCAST_DEBUG
+#endif /* ifndef MEADCAST_DEBUG */
+
+int md_rcv(struct sk_buff *skb)
+{
+    struct ipv6hdr *iphdr;
+    struct ipv6_rt_md_hdr *hdr;
+
+    iphdr = ipv6_hdr(skb);
+
+    printk("MEADcast: Got a packet from %pI6c\n", &iphdr->saddr);
+
+    hdr = (struct ipv6_rt_md_hdr *)skb_transport_header(skb);
+
+    if (hdr->rt_md_type != IPV6_MEADCAST) {
+        printk("MEADcast: early exit\n");
+        return -1;
+    }
+
+    // TODO: handle wrong num_dst
+
+    // if (hdr->discovery) {
+    //     printk("MEADcast: Its a discovery packet\n");
+    //     // forward packet
+    //     // respond sender
+    // } else {
+    //     printk("MEADcast: Its a data packet\n");
+    // }
+
+    #ifdef MEADCAST_DEBUG
+    print_debug(iphdr, hdr);
+    #endif /* ifdef MEADCAST_DEBUG */
+
+    for (int i = 0; i < hdr->num_dst; i++)
+        printk("MEADcast: Dst[%d] is %pI6c\n", i, &hdr->dst[i]);
+
+    return 0;
+}
+
+int md_dcv_rcv(struct sk_buff *skb)
+{
+    struct ipv6hdr *iphdr;
+    struct ipv6_rt_md_hdr *hdr;
+
+    iphdr = ipv6_hdr(skb);
+    hdr = (struct ipv6_rt_md_hdr *)skb_transport_header(skb);
+
+    if (hdr->rt_md_type != IPV6_MEADCAST) {
+        printk("MEADcast: early exit\n");
+        return -1;
+    }
+
+    printk("MEADcast: Got a packet from %pI6c to %pI6c\n", &iphdr->saddr, &iphdr->daddr);
+    #ifdef MEADCAST_DEBUG
+    print_debug(iphdr, hdr);
+    #endif /* ifdef MEADCAST_DEBUG */
+
+    hdr->hopcount++;
+    dcv_rsp(skb);
+
+    return 0;
+}
+
+void print_debug(struct ipv6hdr *iphdr, struct ipv6_rt_md_hdr *hdr)
+{
+    printk("MEADcast:\n"
+           "  num dst:\t%u\n"
+           "  discovery:\t%d\n"
+           "  response:\t%d\n"
+           "  hopcount:\t%d\n"
+           "  dlvmap:\t%u\n"
+           "  rtmap:\t%u\n",
+           hdr->num_dst, hdr->discovery ? 1 : 0, hdr->response ? 1 : 0,
+           hdr->hopcount, ntohl(hdr->dlvmap), ntohl(hdr->rtmap));
+}
+
+int dcv_rsp(struct sk_buff *skb)
+{
+    struct sk_buff *rsp;
+    struct sock *sk;
+    struct ipv6hdr *oip, *rsp_iphdr;
+    struct ipv6_rt_md_hdr *rsp_hdr;
+    struct net *net;
+    struct net_device *in_netdev;
+    struct dst_entry *dst;
+    struct flowi6 fl6;
+
+    // clone input packet for discovery response
+    rsp = skb_copy(skb, GFP_ATOMIC);
+
+    if (!rsp)
+        return 1;
+
+    rsp_iphdr = ipv6_hdr(rsp);
+    rsp_hdr = (struct ipv6_rt_md_hdr *)skb_transport_header(rsp);
+
+    rsp_iphdr->daddr = rsp_iphdr->saddr;
+    rsp_iphdr->hop_limit = 255;
+
+    if (!skb->dev)
+        return 1;
+    net = dev_net(rsp->dev);
+    sk = rsp->sk;
+
+    memset(&fl6, 0, sizeof(fl6));
+    fl6.daddr = rsp_iphdr->daddr;
+    fl6.flowi6_oif = skb->dev->ifindex;
+
+    // get ip of interface
+    in_netdev = dev_get_by_index(net, IP6CB(skb)->iif);
+    if (!in_netdev) {
+        printk("Couldn't get input interface\n");
+        return 1;
+    }
+
+    ipv6_dev_get_saddr(net, in_netdev, &fl6.daddr,
+                sk ? inet6_sk(sk)->srcprefs : 0,
+                &fl6.saddr);
+    dev_put(in_netdev);
+    rsp_iphdr->saddr = fl6.saddr;
+
+    oip = ipv6_hdr(skb);
+    printk("(old) src: %pI6c, dst: %pI6c\n", &oip->saddr, &oip->daddr);
+    printk("(new) src: %pI6c, dst: %pI6c\n", &rsp_iphdr->saddr, &rsp_iphdr->daddr);
+
+    dst = ip6_route_output(net, sk, &fl6);
+    if (IS_ERR(dst)) {
+        printk("Error during route lookup!\n");
+        return 1;
+    }
+
+    return dst_output(net, rsp->sk, rsp);
+}
