diff --git a/include/linux/mdcast.h b/include/linux/mdcast.h
new file mode 100644
index 000000000000..f8b88b0c41f0
--- /dev/null
+++ b/include/linux/mdcast.h
@@ -0,0 +1,47 @@
+#ifndef _LINUX_MDCAST_H
+#define _LINUX_MDCAST_H
+
+#include "linux/in6.h"
+#include "linux/ipv6.h"
+#include "linux/stddef.h"
+#include <linux/ipv6.h>
+#include <linux/skbuff.h>
+
+int md_rcv(struct sk_buff *skb);
+int md_dcv_rcv(struct sk_buff *skb);
+int md_data_rcv(struct sk_buff *skb);
+int dcv_rsp(struct sk_buff *skb);
+
+/*
+ * May using similar approach as rpl.h::ipv6_rpl_sr_hdr.
+ * For example, segments property is used in exthdrs.c:585
+ */
+
+/*
+ * Fix sized part of MEADcast header, bc so far I'm not sure how to handle two
+ * var sized properties in struct (dst array and router map)
+ * substitute hop-count by segments left
+ *  8 bit num dsts
+ *  1 bit disc flag
+ *  1 bit resp flag
+ * 30 bit reserved
+ */
+
+struct ipv6_rt_md_hdr {
+    struct ipv6_rt_hdr rt_hdr;
+    __u32
+        num_dst:    8,
+        discovery:  1,
+        response:   1,
+        hopcount:   6,
+        reserved:  16;
+    __u32 dlvmap;
+    __u32 rtmap;
+    struct in6_addr dst[];
+
+#define rt_md_type		rt_hdr.type
+} __attribute__((packed));
+
+void print_debug(struct ipv6hdr *iphdr, struct ipv6_rt_md_hdr *hdr);
+
+#endif // !_LINUX_MDCAST_H
diff --git a/include/net/ipv6.h b/include/net/ipv6.h
index 2acc4c808d45..7d08915496d0 100644
--- a/include/net/ipv6.h
+++ b/include/net/ipv6.h
@@ -1271,6 +1271,9 @@ static inline int snmp6_unregister_dev(struct inet6_dev *idev) { return 0; }
 #ifdef CONFIG_SYSCTL
 struct ctl_table *ipv6_icmp_sysctl_init(struct net *net);
 struct ctl_table *ipv6_route_sysctl_init(struct net *net);
+#ifdef CONFIG_IPV6_MEADCAST
+struct ctl_table *ipv6_mdcast_sysctl_init(struct net *net);
+#endif
 int ipv6_sysctl_register(void);
 void ipv6_sysctl_unregister(void);
 #endif
diff --git a/include/net/netns/ipv6.h b/include/net/netns/ipv6.h
index 5f2cfd84570a..95d0a894655e 100644
--- a/include/net/netns/ipv6.h
+++ b/include/net/netns/ipv6.h
@@ -19,6 +19,9 @@ struct netns_sysctl_ipv6 {
 	struct ctl_table_header *icmp_hdr;
 	struct ctl_table_header *frags_hdr;
 	struct ctl_table_header *xfrm6_hdr;
+#ifdef CONFIG_IPV6_MEADCAST
+	struct ctl_table_header *mdcast_hdr;
+#endif
 #endif
 	int flush_delay;
 	int ip6_rt_max_size;
@@ -56,6 +59,10 @@ struct netns_sysctl_ipv6 {
 	u8 skip_notify_on_dev_down;
 	u8 fib_notify_on_flag_change;
 	u8 icmpv6_error_anycast_as_unicast;
+#ifdef CONFIG_IPV6_MEADCAST
+    u8 ip6_mdcast_enable;
+    u8 ip6_mdcast_min_dsts;
+#endif
 };
 
 struct netns_ipv6 {
diff --git a/include/uapi/linux/ipv6.h b/include/uapi/linux/ipv6.h
index ac56605fe9bc..8a1d66a979b7 100644
--- a/include/uapi/linux/ipv6.h
+++ b/include/uapi/linux/ipv6.h
@@ -43,6 +43,7 @@ struct in6_ifreq {
 #define IPV6_SRCRT_TYPE_2	2	/* IPv6 type 2 Routing Header	*/
 #define IPV6_SRCRT_TYPE_3	3	/* RPL Segment Routing with IPv6 */
 #define IPV6_SRCRT_TYPE_4	4	/* Segment Routing with IPv6 */
+#define IPV6_MEADCAST 253   /* Using experimental header for MEADcast */
 
 /*
  *	routing header
diff --git a/include/uapi/linux/sysctl.h b/include/uapi/linux/sysctl.h
index 8981f00204db..9775a5d84ae6 100644
--- a/include/uapi/linux/sysctl.h
+++ b/include/uapi/linux/sysctl.h
@@ -530,6 +530,7 @@ enum {
 	NET_IPV6_IP6FRAG_TIME=23,
 	NET_IPV6_IP6FRAG_SECRET_INTERVAL=24,
 	NET_IPV6_MLD_MAX_MSF=25,
+    NET_IPV6_MEADCAST=26,
 };
 
 enum {
@@ -582,6 +583,12 @@ enum {
 	NET_IPV6_ICMP_ECHO_IGNORE_ALL = 2
 };
 
+/* /proc/sys/net/ipv6/meadcast */
+enum {
+    NET_IPV6_MEADCAST_ENABLE = 1,
+    NET_IPV6_MEADCAST_MIN_DSTS = 2
+};
+
 /* /proc/sys/net/<protocol>/neigh/<dev> */
 enum {
 	NET_NEIGH_MCAST_SOLICIT = 1,
diff --git a/net/ipv6/Kconfig b/net/ipv6/Kconfig
index 08d4b7132d4c..809bd04a3476 100644
--- a/net/ipv6/Kconfig
+++ b/net/ipv6/Kconfig
@@ -340,4 +340,13 @@ config IPV6_IOAM6_LWTUNNEL
 
 	  If unsure, say N.
 
+config IPV6_MEADCAST
+	bool "IPv6: MEADcast support"
+	depends on IPV6
+    default y
+	help
+	  Support for MEADcast (Multicast to Explicit Agnostic Destinations).
+
+	  If unsure, say N.
+
 endif # IPV6
diff --git a/net/ipv6/Makefile b/net/ipv6/Makefile
index 3036a45e8a1e..9a45c28b69ec 100644
--- a/net/ipv6/Makefile
+++ b/net/ipv6/Makefile
@@ -26,6 +26,7 @@ ipv6-$(CONFIG_IPV6_SEG6_LWTUNNEL) += seg6_iptunnel.o seg6_local.o
 ipv6-$(CONFIG_IPV6_SEG6_HMAC) += seg6_hmac.o
 ipv6-$(CONFIG_IPV6_RPL_LWTUNNEL) += rpl_iptunnel.o
 ipv6-$(CONFIG_IPV6_IOAM6_LWTUNNEL) += ioam6_iptunnel.o
+ipv6-$(CONFIG_IPV6_MEADCAST) += mdcast.o
 
 obj-$(CONFIG_INET6_AH) += ah6.o
 obj-$(CONFIG_INET6_ESP) += esp6.o
diff --git a/net/ipv6/af_inet6.c b/net/ipv6/af_inet6.c
index 9f9c4b838664..7d34cf6be9d8 100644
--- a/net/ipv6/af_inet6.c
+++ b/net/ipv6/af_inet6.c
@@ -975,6 +975,11 @@ static int __net_init inet6_net_init(struct net *net)
 	net->ipv6.sysctl.ioam6_id = IOAM6_DEFAULT_ID;
 	net->ipv6.sysctl.ioam6_id_wide = IOAM6_DEFAULT_ID_WIDE;
 
+#ifdef CONFIG_IPV6_MEADCAST
+    net->ipv6.sysctl.ip6_mdcast_enable = 1;
+    net->ipv6.sysctl.ip6_mdcast_min_dsts = 1;
+#endif
+
 	err = ipv6_init_mibs(net);
 	if (err)
 		return err;
diff --git a/net/ipv6/exthdrs.c b/net/ipv6/exthdrs.c
index 202fc3aaa83c..dd0338174908 100644
--- a/net/ipv6/exthdrs.c
+++ b/net/ipv6/exthdrs.c
@@ -28,6 +28,10 @@
 #include <linux/slab.h>
 #include <linux/export.h>
 
+#ifdef CONFIG_IPV6_MEADCAST
+#include <linux/mdcast.h>
+#endif /* ifdef CONFIG_IPV6_MEADCAST */
+
 #include <net/dst.h>
 #include <net/sock.h>
 #include <net/snmp.h>
@@ -684,6 +688,10 @@ static int ipv6_rthdr_rcv(struct sk_buff *skb)
 	case IPV6_SRCRT_TYPE_3:
 		/* rpl segment routing */
 		return ipv6_rpl_srh_rcv(skb);
+#ifdef CONFIG_IPV6_MEADCAST
+    case IPV6_MEADCAST:
+        return md_rcv(skb);
+#endif
 	default:
 		break;
 	}
diff --git a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
index 1e8c90e97608..f0ce82d33e6b 100644
--- a/net/ipv6/ip6_output.c
+++ b/net/ipv6/ip6_output.c
@@ -22,6 +22,7 @@
  *				for datagram xmit
  */
 
+#include "linux/mdcast.h"
 #include <linux/errno.h>
 #include <linux/kernel.h>
 #include <linux/string.h>
@@ -647,6 +648,10 @@ int ip6_forward(struct sk_buff *skb)
 
 	hdr->hop_limit--;
 
+    if (hdr->nexthdr == IPPROTO_ROUTING) {
+        md_rcv(skb);
+    }
+
 	return NF_HOOK(NFPROTO_IPV6, NF_INET_FORWARD,
 		       net, NULL, skb, skb->dev, dst->dev,
 		       ip6_forward_finish);
diff --git a/net/ipv6/mdcast.c b/net/ipv6/mdcast.c
new file mode 100644
index 000000000000..311d468ed394
--- /dev/null
+++ b/net/ipv6/mdcast.c
@@ -0,0 +1,383 @@
+#include "asm/checksum_64.h"
+#include "linux/gfp_types.h"
+#include "linux/in.h"
+#include "linux/ipv6.h"
+#include "linux/netdevice.h"
+#include "linux/skbuff.h"
+#include "linux/types.h"
+#include "linux/udp.h"
+#include "net/addrconf.h"
+#include "net/dst.h"
+#include "net/ip6_checksum.h"
+#include "net/ip6_route.h"
+#include "net/udp.h"
+#include <linux/printk.h>
+#include <linux/mdcast.h>
+#include <linux/in.h>
+#include <linux/string.h>
+
+#ifndef INET6_ADDRSTRLEN
+#define INET6_ADDRSTRLEN 46
+#endif
+
+#define CHECK_BIT(var,pos) ((var) & (1<<(pos)))
+
+/* Mapping from lowest index on highest, to ease traversing bitmap backwards.
+ * E.g. GET_I(0) -> 31, GET_I(1) -> 30, ..., GET_I(31) -> 0 */
+#define GET_I(i) (sizeof(((struct ipv6_rt_md_hdr *)0)->rtmap) * 8 - 1 - i)
+
+size_t get_mdc_pkt_size(size_t num_dst)
+{
+    size_t mdc_pl, pad;
+
+    mdc_pl = sizeof(struct ipv6_rt_md_hdr) + num_dst * sizeof(struct in6_addr);
+
+    // num of padding bytes to get a multiple of 8 bytes
+    pad = (8 - (mdc_pl & 7)) & 7;
+
+    return mdc_pl + pad;
+}
+
+int md_rcv(struct sk_buff *skb)
+{
+    struct net *net;
+    struct ipv6hdr *iphdr;
+    struct ipv6_rt_md_hdr *hdr;
+
+    net = dev_net(skb->dev);
+    iphdr = ipv6_hdr(skb);
+    hdr = (struct ipv6_rt_md_hdr *)skb_transport_header(skb);
+
+    if (!net->ipv6.sysctl.ip6_mdcast_enable) {
+        printk("MEADcast: disabled!\n");
+        return 0;
+    }
+
+    if (hdr->rt_md_type != IPV6_MEADCAST) {
+        printk("MEADcast: early exit\n");
+        return -1;
+    }
+
+    // TODO: handle wrong num_dst
+    if (hdr->discovery) {
+        md_dcv_rcv(skb);
+    } else {
+        md_data_rcv(skb);
+    }
+
+    return 0;
+}
+
+int md_dcv_rcv(struct sk_buff *skb)
+{
+    struct ipv6hdr *iphdr;
+    struct ipv6_rt_md_hdr *hdr;
+
+    iphdr = ipv6_hdr(skb);
+    hdr = (struct ipv6_rt_md_hdr *)skb_transport_header(skb);
+
+    if (hdr->rt_md_type != IPV6_MEADCAST ||
+        hdr->rt_hdr.nexthdr != IPPROTO_NONE ||
+        !hdr->discovery || hdr->response ||
+        hdr->num_dst != 1
+    ) {
+        // printk("MEADcast: invalid discovery request!\n");
+        return -1;
+    }
+
+    // printk("MEADcast: Got a packet from %pI6c to %pI6c for %pI6c on %d\n",
+    //        &iphdr->saddr, &iphdr->daddr, &hdr->dst, skb->skb_iif);
+
+    hdr->hopcount++;
+    dcv_rsp(skb);
+
+    return 0;
+}
+
+int set_dst(struct sk_buff *skb, struct in6_addr *daddr)
+{
+    struct flowi6 fl6;
+    struct dst_entry *dst;
+
+    memset(&fl6, 0, sizeof(fl6));
+    fl6.daddr = *daddr;
+    dst = ip6_route_output(dev_net(skb->dev), skb->sk, &fl6);
+
+    if (dst->error) {
+        printk("Error during route lookup of %pI6c", daddr);
+        dst_release(dst);
+        return 1;
+    }
+
+	skb_dst_set(skb, dst);
+    return 0;
+}
+
+int m2u(struct sk_buff *skb, struct in6_addr *dst)
+{
+    struct ipv6hdr ip, *pull;
+    struct ipv6_rt_md_hdr *hdr;
+    struct udphdr *uh;
+    size_t rthdr_len, ip_plen;
+    unsigned int sum;
+
+    // printk("MEADcast M2U of %pI6c", dst);
+    ip  = *(struct ipv6hdr *) skb_network_header(skb);
+    hdr =  (struct ipv6_rt_md_hdr *) skb_transport_header(skb);
+
+    if (!dst)
+        return -1;
+
+    rthdr_len = (hdr->rt_hdr.hdrlen + 1) * 8;
+    ip_plen = ntohs(ip.payload_len) - rthdr_len;
+    uh = (void *) hdr + rthdr_len;
+
+    ip.payload_len = htons(ip_plen);
+    ip.nexthdr = hdr->rt_hdr.nexthdr;
+
+    sum = csum_partial(dst, sizeof(*dst), ~csum_unfold(uh->check));
+    sum = csum_sub(sum, csum_partial(&ip.daddr, sizeof(ip.daddr), 0));
+    uh->check = csum_fold(sum);
+    ip.daddr = *dst;
+
+	if (uh->check == 0)
+		uh->check = CSUM_MANGLED_0;
+
+    pull = skb_pull(skb, rthdr_len);
+    *pull = ip;
+    skb_reset_network_header(skb);
+
+    return set_dst(skb, &ip.daddr);
+}
+
+int md_data_fwd(struct sk_buff *skb, struct net *net, int i, int j)
+{
+    int len, k;
+    unsigned int sum;
+    size_t rthdr_len;
+    struct sk_buff *cpy;
+    struct net *cnet;
+    struct ipv6hdr *ip;
+    struct ipv6_rt_md_hdr *hdr;
+    struct udphdr *uh;
+    struct flowi6 fl6;
+    struct in6_addr *dst;
+
+    ip = (struct ipv6hdr *) skb_network_header(skb);
+    hdr = (struct ipv6_rt_md_hdr *) skb_transport_header(skb);
+
+    len = j - i - 1;
+
+    if (len < net->ipv6.sysctl.ip6_mdcast_min_dsts ||
+        ipv6_chk_addr(net, &hdr->dst[i], skb->dev, 0)
+    ) {
+        for (k = i + 1; k < j - 1; k++) {
+            cpy = pskb_copy(skb, GFP_ATOMIC);
+            cnet = dev_net(cpy->dev);
+            fl6.daddr = hdr->dst[k];
+            // printk("M2U for k=%d ip=%pI6c", k, &hdr->dst[k]);
+            m2u(cpy, &fl6.daddr);
+
+            if (dst_output(cnet, cpy->sk, cpy)) {
+                printk("Error during m2u!\n");
+                return 1;
+            }
+        }
+
+        // printk("M2U if last k=%d ip=%pI6c", k, &hdr->dst[k]);
+        return m2u(skb, &hdr->dst[k]);
+    }
+
+    rthdr_len = (hdr->rt_hdr.hdrlen + 1) * 8;
+    uh = (void *) hdr + rthdr_len;
+    dst = &hdr->dst[i + 1];
+
+    sum = csum_partial(dst, sizeof(*dst), ~csum_unfold(uh->check));
+    sum = csum_sub(sum, csum_partial(&ip->daddr, sizeof(ip->daddr), 0));
+    uh->check = csum_fold(sum);
+
+	if (uh->check == 0)
+		uh->check = CSUM_MANGLED_0;
+
+    hdr->dlvmap = 0 | 1 << GET_I(i);
+    ip->daddr = hdr->dst[i + 1];
+    // printk("M2U last i+1=%d ip=%pI6c", i + 1, &ip->daddr);
+    return set_dst(skb, &ip->daddr);
+}
+
+int md_data_rcv(struct sk_buff *skb)
+{
+    int i, j, n, len, rt, dlv, r1, r2;
+    struct ipv6_rt_md_hdr *hdr;
+    struct sk_buff *cpy;
+    struct net *net;
+
+    hdr = (struct ipv6_rt_md_hdr *) skb_transport_header(skb);
+    if (hdr->discovery || hdr->response) {
+        printk("MEADcast: invalid data packet!\n");
+        return -1;
+    }
+
+    hdr->hopcount++;
+
+    r1 = -1;
+    r2 =  0;
+    n = sizeof(hdr->rtmap) * 8;
+
+    for (i = 0; i < hdr->num_dst; i++) {
+
+        j = GET_I(i);
+        rt  = CHECK_BIT(hdr->rtmap, j);
+        dlv = CHECK_BIT(hdr->dlvmap, j);
+
+        if (!rt)
+            continue;
+
+        if (r1 >= 0) {
+
+            if (!r2)
+                r2 = i;
+
+            /* If r2 has to be delivered we have to create a replica. */
+            if (dlv) {
+
+                len = r2 - r1 - 1;
+                // printk("Deliver %d eps via rt at %d (copy)\n", len, r1);
+
+                cpy = pskb_copy(skb, GFP_ATOMIC);
+                net = dev_net(cpy->dev);
+                md_data_fwd(cpy, net, r1, r2);
+
+                if (dst_output(net, cpy->sk, cpy)) {
+                    printk("Error during md data fwd!\n");
+                    return 1;
+                }
+
+                r1 = i;
+                r2 = 0;
+            }
+
+            continue;
+        }
+
+        if (dlv)
+            r1 = i;
+    }
+
+    /* If r1 is the last router set r2 to the last index. */
+    if (!r2)
+        r2 = i;
+
+    len = r2 - r1 - 1;
+    // printk("Deliver %d eps via rt at %d\n", len, r1);
+    md_data_fwd(skb, dev_net(skb->dev), r1, r2);
+
+    // printk("FIN: %s\n", le);
+    return 0;
+}
+
+int dcv_rsp(struct sk_buff *skb)
+{
+    struct sk_buff *rsp;
+    struct net *rnet;
+    struct ipv6hdr *rip;
+    struct ipv6_rt_md_hdr *rhdr;
+    struct flowi6 fl6;
+    struct dst_entry *dst;
+
+    // clone input packet for discovery response
+    rsp = pskb_copy(skb, GFP_ATOMIC); 
+
+    if (!rsp || !skb->dev)
+        return 1;
+
+    rnet = dev_net(rsp->dev);
+
+    rip = ipv6_hdr(rsp);
+    rip->daddr = rip->saddr;
+    rip->hop_limit = 255;
+
+    rhdr = (struct ipv6_rt_md_hdr *)skb_transport_header(rsp);
+    rhdr->response = 1;
+
+    memset(&fl6, 0, sizeof(fl6));
+    fl6.daddr = rip->daddr;
+
+    // get route back to sender
+    dst = ip6_route_output(rnet, rsp->sk, &fl6);
+    if (dst->error) {
+        printk("Error during route lookup\n!");
+        dst_release(dst);
+        return 1;
+    }
+
+	// skb_dst_drop(rsp);
+	skb_dst_set(rsp, dst);
+
+    // get ip of net_device
+    if (ipv6_dev_get_saddr(rnet, dst->dev, &fl6.daddr,
+                rsp->sk ? inet6_sk(rsp->sk)->srcprefs : 0,
+                &fl6.saddr)) {
+        printk("Error during if saddr lookup!\n");
+        dev_put(dst->dev);
+        return 1;
+    }
+    dev_put(dst->dev);
+    rip->saddr = fl6.saddr;
+
+    if (dst_output(rnet, rsp->sk, rsp)) {
+        printk("Error during md response!\n");
+        return 1;
+    }
+
+    return 0;
+}
+
+void print_debug(struct ipv6hdr *iphdr, struct ipv6_rt_md_hdr *hdr)
+{
+    printk("MEADcast:\n"
+           "  num dst:\t%u\n"
+           "  discovery:\t%d\n"
+           "  response:\t%d\n"
+           "  hopcount:\t%d\n"
+           "  dlvmap:\t%u\n"
+           "  rtmap:\t%u\n",
+           hdr->num_dst, hdr->discovery ? 1 : 0, hdr->response ? 1 : 0,
+           hdr->hopcount, ntohl(hdr->dlvmap), ntohl(hdr->rtmap));
+}
+
+#ifdef CONFIG_SYSCTL
+static struct ctl_table ipv6_mdcast_table_template[] = {
+    {
+		.procname	= "enable",
+		.data		= &init_net.ipv6.sysctl.ip6_mdcast_enable,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= proc_dou8vec_minmax,
+    },
+    {
+		.procname	= "min_dsts",
+		.data		= &init_net.ipv6.sysctl.ip6_mdcast_min_dsts,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= proc_dou8vec_minmax,
+    },
+    { },
+};
+
+struct ctl_table * __net_init ipv6_mdcast_sysctl_init(struct net *net)
+{
+    struct ctl_table *table;
+
+    table = kmemdup(ipv6_mdcast_table_template,
+                    sizeof(ipv6_mdcast_table_template),
+                    GFP_KERNEL);
+
+    if (table) {
+        table[0].data = &net->ipv6.sysctl.ip6_mdcast_enable;
+        table[1].data = &net->ipv6.sysctl.ip6_mdcast_min_dsts;
+    }
+
+    return table;
+}
+#endif
diff --git a/net/ipv6/sysctl_net_ipv6.c b/net/ipv6/sysctl_net_ipv6.c
index 94a0a294c6a1..146c02985b93 100644
--- a/net/ipv6/sysctl_net_ipv6.c
+++ b/net/ipv6/sysctl_net_ipv6.c
@@ -6,6 +6,7 @@
  * YOSHIFUJI Hideaki @USAGI:	added icmp sysctl table.
  */
 
+#include "net/net_namespace.h"
 #include <linux/mm.h>
 #include <linux/sysctl.h>
 #include <linux/in6.h>
@@ -289,9 +290,28 @@ static int __net_init ipv6_sysctl_net_init(struct net *net)
 	if (!net->ipv6.sysctl.icmp_hdr)
 		goto out_unregister_route_table;
 
+#ifdef CONFIG_IPV6_MEADCAST
+	struct ctl_table *ipv6_mdcast_table;
+
+    ipv6_mdcast_table = ipv6_mdcast_sysctl_init(net);
+	if (!ipv6_mdcast_table)
+        goto out_ipv6_mdcast_table;
+
+    net->ipv6.sysctl.mdcast_hdr =
+        register_net_sysctl(net, "net/ipv6/meadcast", ipv6_mdcast_table);
+    if (!net->ipv6.sysctl.mdcast_hdr)
+        goto out_unregister_mdcast_table;
+#endif
+
 	err = 0;
 out:
 	return err;
+#ifdef CONFIG_IPV6_MEADCAST
+out_unregister_mdcast_table:
+    unregister_net_sysctl_table(net->ipv6.sysctl.mdcast_hdr);
+out_ipv6_mdcast_table:
+    kfree(ipv6_mdcast_table);
+#endif
 out_unregister_route_table:
 	unregister_net_sysctl_table(net->ipv6.sysctl.route_hdr);
 out_unregister_ipv6_table:
@@ -322,6 +342,13 @@ static void __net_exit ipv6_sysctl_net_exit(struct net *net)
 	kfree(ipv6_table);
 	kfree(ipv6_route_table);
 	kfree(ipv6_icmp_table);
+
+#ifdef CONFIG_IPV6_MEADCAST
+	struct ctl_table *ipv6_mdcast_table;
+	ipv6_mdcast_table = net->ipv6.sysctl.mdcast_hdr->ctl_table_arg;
+	unregister_net_sysctl_table(net->ipv6.sysctl.hdr);
+	kfree(ipv6_icmp_table);
+#endif
 }
 
 static struct pernet_operations ipv6_sysctl_net_ops = {
